core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -540,9 +540,11 @@ private[spark] class ExecutorAllocationManager(
 
     private val stageIdToNumTasks = new mutable.HashMap[Int, Int]
     private val stageIdToTaskIndices = new mutable.HashMap[Int, mutable.HashSet[Int]]
-    private val executorIdToTaskIds = new mutable.HashMap[String, mutable.HashSet[Long]]
-    // Number of tasks currently running on the cluster.  Should be 0 when no stages are active.
-    private var numRunningTasks: Int = _
+    private val executorIdToStageAndNumTasks =
+      new mutable.HashMap[String, mutable.HashMap[Int, Int]]
+
+    // Track stage id and related running tasks.
+    private val stageIdToRunningTasks = new mutable.HashMap[Int, mutable.HashSet[Int]]
```

Comment:
this needs to specify task index. Right now it's not clear whether the values are task indices or task IDs.